opm/common/util/numeric/cmp.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+  Copyright 2016 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <math.h>
+#include <string.h>
+
+#include <cstddef>
+#include <vector>
+
+#include <opm/common/util/numeric/cmp.hpp>
+
+/**
+   Ahhh - the joys of comparing floating point numbers ....
+
+   http://www.cygnus-software.com/papers/comparingfloats/comparingfloats.htm
+*/
+
+
+namespace Opm {
+    namespace cmp {
+
+        bool double_equal(double value1 , double value2 , double abs_eps , double rel_eps) {
+            bool equal = true;
+            double diff = fabs(value1 - value2);
+            if (diff > abs_eps) {
+                double sum = fabs(value1) + fabs(value2);
+
+                if (diff > sum * rel_eps) {
+                    equal = false;
+                }
+            }
+            return equal;
+        }
+
+
+        bool double_equal(double value1 , double value2) {
+            return double_equal( value1 , value2 , abs_epsilon , rel_epsilon );
+        }
+
+        /*****************************************************************/
+
+        bool double_vector_equal(const std::vector<double>& v1,const std::vector<double>& v2, double abs_eps , double rel_eps) {
+            if (v1.size() != v2.size()) {
+                return false;
+            }
+
+            for (size_t i = 0; i < v1.size(); i++) {
+                if (!double_equal( v1[i] , v2[i]))
+                    return false;
+            }
+
+            return true;
+        }
+
+        bool double_vector_equal(const std::vector<double>& v1,const std::vector<double>& v2) {
+            return double_vector_equal( v1 , v2 , abs_epsilon , rel_epsilon);
+        }
+
+
+        /*
+          This function is inspired by memcmp(); it will return 0 if
+          the two memory locations are 'equal enough'. If the memory
+          locations are not equal it will return 1 - no attempt to
+          differentiate between p1 < p2 or p1 > p2.
+        */
+        int double_memcmp(const double * p1 , const double *p2 , size_t num_elements, double abs_eps , double rel_eps) {
+            if (memcmp(p1 , p2 , num_elements * sizeof * p1) == 0)
```

Comment:
I think it would be clearer to write sizeof(*p1), unless that is wrong or unidiomatic for some reason (I do not claim to be a C programmer...).